rmw_connext_cpp/src/rmw_init.cpp

```diff
@@ -51,20 +57,38 @@ rmw_init_options_copy(const rmw_init_options_t * src, rmw_init_options_t * dst)
     RMW_SET_ERROR_MSG("expected zero-initialized dst");
     return RMW_RET_INVALID_ARGUMENT;
   }
+  const rcutils_allocator_t * allocator = &src->allocator;
```

Comment:
I can use a reference instead of a pointer.
About constness, we've some functions that take a [const pointer](https://github.com/ros2/rcutils/blob/af3e14543ffa7e8d4d9d38f9c242548f76accd89/include/rcutils/types/array_list.h#L126), there are others taking it [by value](https://github.com/ros2/rcutils/blob/af3e14543ffa7e8d4d9d38f9c242548f76accd89/include/rcutils/types/string_map.h#L99), and others taking a [mutable pointer](https://github.com/ros2/rmw/blob/f76511333709a6f5d1e4543129ec73f2cd2b4f06/rmw/include/rmw/names_and_types.h#L72).

In the first and last case, you can do:
```cpp
allocator->allocate(size, allocator->state);
```
In the second case, you can replace `->` with `.`.

Why does [this](https://github.com/ros2/rcutils/blob/af3e14543ffa7e8d4d9d38f9c242548f76accd89/src/array_list.c#L71) work? I will have to check, but I think that C structs constness is not propagated to its members. So, it doesn't matter that [allocate](https://github.com/ros2/rcutils/blob/af3e14543ffa7e8d4d9d38f9c242548f76accd89/include/rcutils/allocator.h#L49) takes a mutable pointer to `state`.

Reply:
> I can use a reference instead of a pointer.

I was mostly interested in the rationale behind explicitly introducing an indirection. 

> About constness, we've some functions that take a const pointer, there are others taking it by value, and others taking a mutable pointer.

I'm aware of how allocators are passed in all forms and colors here and there. That's something we can discuss in a follow-up issue though.

> Why does this work? I will have to check, but I think that C structs constness is not propagated to its members. So, it doesn't matter that allocate takes a mutable pointer to state.

A `const` qualified `struct` in C/C++ implies that none of its members can be mutated. In the case you mention, you're actually not mutating `state`, the pointer. AFAIK cv-qualifiers do not propagate through indirections. That's why a `const std::shared_ptr<SomeObject>` will happily let you mutate `SomeObject`, whereas `const std::shared_ptr<const SomeObject>` will likely get you what you want.

Reply:
> I was mostly interested in the rationale behind explicitly introducing an indirection.

Ok, I can use a reference if you want.

> I'm aware of how allocators are passed in all forms and colors here and there. That's something we can discuss in a follow-up issue though.

:+1:

> A const qualified struct in C/C++ implies that none of its members can be mutated. In the case you mention, you're actually not mutating state, the pointer. AFAIK cv-qualifiers do not propagate through indirections. That's why a const std::shared_ptr<SomeObject> will happily let you mutate SomeObject, whereas const std::shared_ptr<const SomeObject> will likely get you what you want.

Yes, exactly. I misunderstood your first comment and derailed a bit.